Packages: cleanups & new distros

thresheek opened the pull request:
This adds Ubuntu 24.04.

While at it, remove some older EOLed distributions from both rpm and deb worlds.

thresheek commented:
A correct version is submitted at https://github.com/nginx/unit/pull/1287

ac000 reviewed (commented):
Hi @thresheek 

```
 Packages: don't redefine FORTIFY_SOURCE on Ubuntu 24+

The default on Ubuntu 24.04 and newer is now -D_FORTIFY_SOURCE=3 which clashes
with our definition.
```

Could you re-flow the commit message to wrap after 72 chars please? That allows for 4 chars of left/right margin in the likes of git-log(1)...

ac000 commented on pkg/deb/debian.module/rules-noarch.in:
```diff
@@ -3,8 +3,18 @@
 #export DH_VERBOSE=1
 
-export DEB_BUILD_MAINT_OPTIONS=hardening=+all,-pie
+include /usr/share/dpkg/vendor.mk
+ifeq ($(shell $(call dpkg_vendor_derives_from,ubuntu)),yes)
+DISTROREL_MAJOR := $(shell lsb_release -rs | cut -d. -f 1)
+ifeq ($(shell test $(DISTROREL_MAJOR) -ge 24; echo $$?), 0)
+export DEB_CFLAGS_MAINT_APPEND=-fPIC
+else
+export DEB_CFLAGS_MAINT_APPEND=-Wp,-D_FORTIFY_SOURCE=2 -fPIC
+endif
+else
 export DEB_CFLAGS_MAINT_APPEND=-Wp,-D_FORTIFY_SOURCE=2 -fPIC
+endif
```
Anyway we can loose one of those duplicate branches?, i.e only having one of

```
export DEB_CFLAGS_MAINT_APPEND=-Wp,-D_FORTIFY_SOURCE=2 -fPIC
```

Ditto for the others...

Or do you envisage them diverging at some point?

thresheek replied:
Oh, it looks like we can drop most of the patch for Ubuntus, since they seem to be using FORTIFY_SOURCE=2 for all their releases since ages, and moved to FORTIFY_SOURCE=3 for 24.04:

https://salsa.debian.org/toolchain-team/gcc/-/commit/ae01f86e0731d2661946866464939aba79b49da9

thresheek replied:
I've updated the patchset accordingly. Thanks!

callahad commented:
@thresheek While we're at it, can we drop CentOS 7 and 8 (both effectively EOL), as well as Amazon Linux 2?

callahad commented:
Unit can drop RHEL 8, too - its Full Support Phase ended last month. 😉

callahad commented:
Less flippantly, dropping CentOS / RHEL 8 moves our baseline GCC from 7 to 9.

Then to GCC 10 when we drop Ubuntu 20.04, and to GCC 11 when we drop Debian 11. Both of which I'd advocate for doing sooner rather than later, but not necessarily in this PR.

thresheek commented:
One can argue that "Maintenance support" for RHEL 8 is pretty much the usual way of supporting any other Linux distros through their lifetime;  see https://access.redhat.com/support/policy/updates/errata#Full_Support_Phase (and further paragraphs) for more details on that.

In general, we'd consider any official non-gated (GA, not additionally paid) support that includes security fixes and errata as something we can also build packages for.   For RHEL, that includes Full Support and Maintenance Support.

Of course this is not my call to make, I'm just adding context on why it's the way it is - and it's your call to make to end the distribution of binary packages for those targets.

callahad commented:
Appreciate that :)

For purposes of keeping this PR on track, let's drop CentOS 7 and leave the rest.

I'll start a separate discussion to define our baseline binary support targets, which we can address in a subsequent PR. (The tl;dr of my position is that we're a tiny team that needs to skew more towards innovation than maintenance; we're currently losing more by imposing these limits than we gain.)

thresheek commented:
I've removed CentOS 7 / RHEL 7 in the new patches.  Looks like we had some more leftovers from Fedora cleanups, and *ahem* CentOS 6 days, so I've removed those as well.

callahad commented:
Here's the separate discussion on packaging targets: https://github.com/nginx/unit/discussions/1333

thresheek commented:
Ping for a review

ac000 commented:
No Fedora 40?

thresheek commented:
It looks like Fedora 40 doesnt require any changes compared to what we have for Fedora 39... And on my VMs it builds the packaging just fine.

lchastel commented:
When do you plan to merge and release this update ?

callahad commented:
Merging should happen shortly. I'd expect us to release Unit 1.33 in the next 6 weeks.
